include/phasar/Utils/Logger.h

```diff
@@ -45,45 +43,131 @@ enum SeverityLevel {
-std::ostream &operator<<(std::ostream &OS, const SeverityLevel &Level);
+class Logger final {
+public:
+  /**
+   * Set the filter level.
+   */
+  static void setLoggerFilterLevel(SeverityLevel Level);
+
+  static SeverityLevel getLoggerFilterLevel();
+
+  static bool isLoggingEnabled();
+
+  static void enable();
+
+  static void disable();
+
+  static llvm::raw_ostream &
+  getLogStream(const std::optional<SeverityLevel> &Level,
+               const std::optional<llvm::StringRef> &Category);
+
+  static bool logCategory(const llvm::StringRef &Category,
+                          const std::optional<SeverityLevel> &Level);
+
+  static void addLinePrefix(llvm::raw_ostream &,
+                            const std::optional<SeverityLevel> &Level,
+                            const std::optional<std::string> &Category);
+
+  static void initializeStdoutLogger(
+      const std::optional<SeverityLevel> &Level = std::nullopt,
+      const std::optional<std::string> &Category = std::nullopt);
+
+  static void initializeStderrLogger(
+      const std::optional<SeverityLevel> &Level = std::nullopt,
+      const std::optional<std::string> &Category = std::nullopt);
+
+  [[nodiscard]] static bool initializeFileLogger(
+      const llvm::StringRef &Filename,
+      const std::optional<SeverityLevel> &Level = std::nullopt,
+      const std::optional<std::string> &Category = std::nullopt,
+      bool Append = false);
+
+private:
+  enum class StdStream : uint8_t { STDOUT = 0, STDERR };
+  static inline bool LoggingEnabled = false;
+  static inline llvm::StringMap<std::map<std::optional<SeverityLevel>,
+                                         std::variant<StdStream, std::string>>>
+      CategoriesToStreamVariant;
+  static inline std::map<std::optional<SeverityLevel>,
+                         std::variant<StdStream, std::string>>
+      LevelsToStreamVariant;
+  static inline SeverityLevel LogFilterLevel = DEBUG;
+  static inline llvm::StringMap<std::unique_ptr<llvm::raw_fd_ostream>>
+      LogfileStreams;
+  // static inline auto StartTime = std::chrono::steady_clock::now();
+  [[nodiscard]] static llvm::raw_ostream &
+  getLogStream(const std::optional<SeverityLevel> &Level,
+               const std::map<std::optional<SeverityLevel>,
+                              std::variant<StdStream, std::string>>
+                   &PassedLevelsToStreamVariant);
+  [[nodiscard]] static llvm::raw_ostream &getLogStreamFromStreamVariant(
+      const std::variant<StdStream, std::string> &StreamVariant);
+};
 
-extern SeverityLevel LogFilterLevel;
+std::string toString(const SeverityLevel &Level);
 
 #ifdef DYNAMIC_LOG
-BOOST_LOG_INLINE_GLOBAL_LOGGER_DEFAULT(
-    lg, boost::log::sources::severity_logger<SeverityLevel>)
-// For performance reason, we want to disable any formatting computation
-// that would go straight into logs if logs are deactivated
-// This macro does just that
-#define LOG_IF_ENABLE_BOOL(condition, computation)                             \
+
```

Comment:
Is the `DYNAMIC_LOG` options still necessary? What's the reasoning behind it?

Reply:
The original reason for introducint it was, that even if logging is dynamically disabled, each log-statement must at least check whether logging is enabled which might have a cost at runtime. The option DYNAMIC_LOG can then disable logging already at compile-time

Reply:
Ah, so you want to have an option to disable this check at all.
I think, we should keep this for now but maybe at some pointer later on, find a 'better' implementation maybe?
Just pasting the macro everywhere, always looks so bloated. My idea goes in the direction of using a `class` with some `consteval` magic that we can use to define the log scopes.
But anyhow, we should discuss this at some other point :D